### DISCUSSION TITLE
Custom Components - Writing to Local Database

### DISCUSSION THREAD

**Discussion by josh99smith** [Budibase/budibase#10517] [Help] [answered]
I am working on a form filling PDF plugin for Budibase. I am having trouble figuring out how to write data to the local budibase database via the Component.svelte script. I read the Budibase docs and they explain that there is an SDK context you can use to work with the local database.

The SDK context gives you access to the internal Budibase API. I have added this to my SVELTE script.
`const { styleable, API } = getContext("sdk")`

However, I can not figure out how to actually use this to write data to the database. I am trying to write data to local table called documentUploads. I am using the following Javascript to write the data into the table. Am I missing something?

```
download(pdfBytes, "pdf-lib_form_creation_example.pdf", "application/pdf");
    const pdfBase64 = btoa("pdf data");
   ###  const uploadsTable = sdk.database.table("ta_ee2d542d9c174941bc46bd0a3abba290");
    const newRecord = {
      DocumentDesc: "Test",
      "Document Attachment": pdfBase64,
    };
    await usersTable.insert(newRecord);
  }
```

When I run this in my app, I get the following error in the console. Which is on line 3 above.

`Uncaught (in promise) ReferenceError: sdk is not defined at HTMLButtonElement.<anonymous> (Component.svelte:45:26)`

**Comment by ConorWebb96** [comment, marked as answer]
Hey @josh99smith,

I wasn't sure whether you wanted to save directly using the API SDK or via a form input approach so I've supplied answers to both which will hopefully aid you.

**Form apporach**

I would suggest looking at other custom plugins which tie into form inputs. https://github.com/Budibase/plugins a good one to look at would like be: https://github.com/chungchunwang/Budibase-File-Upload as its doing file upload as well.

From memory I believe you would need to to also include the form contexts for the above to get it working within a form. I've taken an example of this directly from the Budibase codebase.

```
  const formContext = getContext("form")
  const formStepContext = getContext("form-step")
  const fieldGroupContext = getContext("field-group")
  const formApi = formContext?.formApi;
  const labelPos = fieldGroupContext?.labelPosition || "above";
  $: formStep = formStepContext ? $formStepContext || 1 : 1
  $: formField = formApi?.registerField(
    field,
    type,
    defaultValue,
    disabled,
    validation,
    formStep
  )
  $: unsubscribe = formField?.subscribe(value => {
    fieldState = value?.fieldState
    fieldApi = value?.fieldApi
  })
  onDestroy(() => {
    fieldApi?.deregister()
    unsubscribe?.()
  })
```

**API approach**

I can see you have included the API SDK correctly. All you need to do is something similar to the below, make sure to include the tableID you are saving to and any fields that may be required and it should save.

```
      await API.saveRow({
        tableId: "ta_ee2d542d9c174941bc46bd0a3abba290",
        <requiredFieldsHere> : <Value here>
      });
```

I'm not quite sure what you are doing with line 3, but hopefully the code provided above helps point you in the right direction. Additionally this custom component has a very good example of saving a row through the API on line 63: https://github.com/aptkingston/budibase-comment-box/blob/main/src/Component.svelte

Hopefully, one of the two answers here work for you. I hope this helps!

**Reply by josh99smith** [reply]
So greatful for your response my friend! I am going to give this a try today. I am sure this will solve my issue, if not point me in the correct direction. Much grattitute!

**Reply by josh99smith** [reply]
Worked like a charm! I am using the API SKD method. I was able to write to fields with use using your help! One last questions, do you know it is possible to send data to an attachment field? I am trying to send a PDF in Base64 to an attachment field, but I can not sort out how the JSON should be structured. This how my saveRow method is structured now.

```
 await API.saveRow({
      tableId: "ta_ee2d542d9c174941bc46bd0a3abba290",
      DocReqID: "test",
      DocumentDesc: "Test2",
      Document_Attachment: [
        {
          data: pdfBase64,
        },
      ],
    });
```
